File: web/src/protocol.ts
/**
 * Signaling protocol types - mirrors the Rust `protocol.rs` definitions.
 *
 * All messages are JSON-serialized and exchanged over WebSocket.
 * The `type` field is snake_case and matches the Rust serde(tag = "type") convention.
 */

export type SdpRole = 'offerer' | 'answerer';
export type IceMode = 'full' | 'lite';
export type DtlsRole = 'active' | 'passive' | 'auto';

export interface SessionConfig {
  client_sdp_role: SdpRole;
  server_ice_mode: IceMode;
  client_dtls_role: DtlsRole;
}

export interface CreateMessage {
  type: 'create';
  session_id: string;
  config: SessionConfig;
}

export interface DestroyMessage {
  type: 'destroy';
  session_id: string;
}

export interface SdpMessage {
  type: 'sdp';
  session_id: string;
  sdp: string;
}

export interface ReadyMessage {
  type: 'ready';
  session_id: string;
}

export type ClientMessage = CreateMessage | DestroyMessage | SdpMessage | ReadyMessage;

export interface CreatedMessage {
  type: 'created';
  session_id: string;
}

export interface DestroyedMessage {
  type: 'destroyed';
  session_id: string;
}

export interface ServerSdpMessage {
  type: 'sdp';
  session_id: string;
  sdp: string;
}

export interface ServerReadyMessage {
  type: 'ready';
  session_id: string;
}

export interface ServerErrorMessage {
  type: 'error';
  session_id: string | null;
  message: string;
}

export type ServerMessage =
  | CreatedMessage
  | DestroyedMessage
  | ServerSdpMessage
  | ServerReadyMessage
  | ServerErrorMessage;

File: src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SdpRole {
    Offerer,
    Answerer,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum IceMode {
    Full,
    Lite,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DtlsRole {
    Active,
    Passive,
    Auto,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionConfig {
    pub client_sdp_role: SdpRole,
    pub server_ice_mode: IceMode,
    pub client_dtls_role: DtlsRole,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    Create {
        session_id: String,
        config: SessionConfig,
    },
    Destroy {
        session_id: String,
    },
    Sdp {
        session_id: String,
        sdp: String,
    },
    Ready {
        session_id: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    Created {
        session_id: String,
    },
    Destroyed {
        session_id: String,
    },
    Sdp {
        session_id: String,
        sdp: String,
    },
    Ready {
        session_id: String,
    },
    Error {
        session_id: Option<String>,
        message: String,
    },
}
